From a89be5ce33f291a350c55dd0c5916eb853d058b1 Mon Sep 17 00:00:00 2001
From: Nick Clark <nicholas.clark@gmail.com>
Date: Wed, 5 Aug 2015 15:39:23 -0700
Subject: [opkg-0.3.0 PATCH 4/4] opkg/libopkg: Added config-file option to
 ignore UIDs
To: opkg-devel@googlegroups.com
Cc: paul@paulbarker.me.uk

On filesystems where UIDs aren't well-behaved, UIDs/GIDs can't
always be trusted to be restorable. Additionally, non-privileged
users who want to use OPKG to install files in some home directory
won't generally be able to control the permissions of the output.

OPKG's default behaviour when faced with a UID mismatch is to report
the warning that libarchive produces, but to continue anyways.

This patch doesn't change the default behavior, but it adds an
'ignore_uid' boolean option to opkg_conf. If enabled in the config
file by 'option ignore_uid', libarchive will be called without the
'ARCHIVE_EXTRACT_OWNER' flag (and UID/GID restoration won't be
attempted).

Signed-off-by: Nick Clark <nicholas.clark@gmail.com>
---
 libopkg/opkg_archive.c | 4 ++++
 libopkg/opkg_conf.c    | 1 +
 libopkg/opkg_conf.h    | 1 +
 3 files changed, 6 insertions(+)

diff --git a/libopkg/opkg_archive.c b/libopkg/opkg_archive.c
index be903e4..2d119b3 100644
--- a/libopkg/opkg_archive.c
+++ b/libopkg/opkg_archive.c
@@ -22,6 +22,7 @@
 #include <stdio.h>
 #include <string.h>
 
+#include "opkg_conf.h"
 #include "opkg_message.h"
 #include "opkg_archive.h"
 #include "file_util.h"
@@ -716,6 +717,9 @@ struct opkg_ar *ar_open_pkg_data_archive(const char *filename)
     ar->extract_flags = ARCHIVE_EXTRACT_OWNER | ARCHIVE_EXTRACT_PERM |
         ARCHIVE_EXTRACT_TIME | ARCHIVE_EXTRACT_UNLINK;
 
+    if (opkg_config->ignore_uid)
+        ar->extract_flags &= ~ARCHIVE_EXTRACT_OWNER;
+
     return ar;
 }
 
diff --git a/libopkg/opkg_conf.c b/libopkg/opkg_conf.c
index 2fd3112..599279d 100644
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -55,6 +55,7 @@ static opkg_option_t options[] = {
     {"force_defaults", OPKG_OPT_TYPE_BOOL, &_conf.force_defaults},
     {"force_maintainer", OPKG_OPT_TYPE_BOOL, &_conf.force_maintainer},
     {"ignore_maintainer", OPKG_OPT_TYPE_BOOL, &_conf.ignore_maintainer},
+    {"ignore_uid", OPKG_OPT_TYPE_BOOL, &_conf.ignore_uid},
     {"force_depends", OPKG_OPT_TYPE_BOOL, &_conf.force_depends},
     {"force_overwrite", OPKG_OPT_TYPE_BOOL, &_conf.force_overwrite},
     {"force_downgrade", OPKG_OPT_TYPE_BOOL, &_conf.force_downgrade},
diff --git a/libopkg/opkg_conf.h b/libopkg/opkg_conf.h
index 13bf124..7bca948 100644
--- a/libopkg/opkg_conf.h
+++ b/libopkg/opkg_conf.h
@@ -78,6 +78,7 @@ typedef struct opkg_conf {
 
     /* options */
     int autoremove;
+    int ignore_uid;
     int force_depends;
     int force_defaults;
     int force_maintainer;
-- 
2.4.5

